File: C++/prim_algorithm.cpp
#include <iostream>
#include <vector>
#include <climits>

using namespace std;

// Utility function to find the vertex with the minimum key value
int minKey(vector<int>& key, vector<bool>& includedSet, int V) {
    int min = INT_MAX, minIndex;
    for (int v = 0; v < V; v++) {
        if (!includedSet[v] && key[v] < min) {
            min = key[v];
            minIndex = v;
        }
    }
    return minIndex;
}

// Prim's Algorithm to find MST
void primMST(vector<vector<int>>& graph, int V) {
    vector<int> parent(V, -1); // Array to store the MST
    vector<int> key(V, INT_MAX); // Key values to pick the minimum weight edge
    vector<bool> includedSet(V, false); // To track vertices included in MST
    
    key[0] = 0;  // Start with the first vertex
    
    for (int count = 0; count < V - 1; count++) {
        int u = minKey(key, includedSet, V);
        includedSet[u] = true;
        
        // Update key values and parent of adjacent vertices
        for (int v = 0; v < V; v++) {
            if (graph[u][v] && !includedSet[v] && graph[u][v] < key[v]) {
                key[v] = graph[u][v];
                parent[v] = u;
            }
        }
    }
    
    // Print the resulting MST
    cout << "Edge \tWeight\n";
    for (int i = 1; i < V; i++) {
        cout << parent[i] << " - " << i << "\t" << graph[i][parent[i]] << endl;
    }
}

int main() {
    int V = 5;
    vector<vector<int>> graph = {
        {0, 2, 0, 6, 0},
        {2, 0, 3, 8, 5},
        {0, 3, 0, 0, 7},
        {6, 8, 0, 0, 9},
        {0, 5, 7, 9, 0}
    };
    
    primMST(graph, V);
    return 0;
}

File: Python/prim_algorithm.py
import sys

# Function to find the vertex with the minimum key value
def min_key(key, included_set, V):
    min_val = sys.maxsize
    min_index = -1
    for v in range(V):
        if not included_set[v] and key[v] < min_val:
            min_val = key[v]
            min_index = v
    return min_index

# Prim's Algorithm to find MST
def prim_mst(graph, V):
    parent = [-1] * V  # Array to store the MST
    key = [sys.maxsize] * V  # Key values to pick the minimum weight edge
    included_set = [False] * V  # To track vertices included in MST
    
    key[0] = 0  # Start with the first vertex
    
    for _ in range(V - 1):
        u = min_key(key, included_set, V)
        included_set[u] = True
        
        # Update key values and parent of adjacent vertices
        for v in range(V):
            if graph[u][v] and not included_set[v] and graph[u][v] < key[v]:
                key[v] = graph[u][v]
                parent[v] = u
    
    # Print the resulting MST
    print("Edge \tWeight")
    for i in range(1, V):
        print(f"{parent[i]} - {i} \t{graph[i][parent[i]]}")

# Example Usage
graph = [
    [0, 2, 0, 6, 0],
    [2, 0, 3, 8, 5],
    [0, 3, 0, 0, 7],
    [6, 8, 0, 0, 9],
    [0, 5, 7, 9, 0]
]

V = 5
prim_mst(graph, V)
